chef/cookbooks/crowbar/recipes/crowbar-upgrade.rb

```diff
@@ -0,0 +1,113 @@
+#
+# Cookbook Name:: crowbar
+# Recipe:: crowbar-upgrade
+#
+# Copyright 2013-2016, SUSE LINUX Products GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#
+# This recipe is for executing actions that need to be done at nodes
+# as a preparation for the system upgrade. The preparation itself
+# consists of several steps which we distinguish by various attributes
+# saved in the node structure.
+
+return unless node[:platform] == "suse"
+
+if node["crowbar_wall"]["crowbar_openstack_upgrade"]
+
+  # Actions to be run last, when admin node is already new SUSE Cloud version (6)
+  # Nodes will be restarted and their system upgraded after this.
+
+  # put HA node into maintenance mode (TODO is it needed? It wasn't in 4-5 upgrade...)
+
+  # Stop pacemaker resources in all clusters
+  bash "stop pacemaker resources" do
+    code <<-EOF
+      for type in clone ms primitive; do
+        for resource in $(crm configure show | grep ^$type | cut -d " " -f2); do
+          crm resource stop $resource
+        done
+      done
+    EOF
+    only_if { ::File.exist?("/usr/sbin/crm") }
+  end
+
+  # Stop openstack services
+  bash "stop HA and openstack services" do
+    code <<-EOF
+      for i in /etc/init.d/openstack-* \
+               /etc/init.d/openvswitch-switch \
+               /etc/init.d/ovs-usurp-config-* \
+               /etc/init.d/drbd \
```

Comment:
@jsuchome this can work if there are not any other openstack service running in the cluster outside HA. Also before the `pg_dump` we need to wait that the rest of the nodes already stopped the openstack services (HA and no HA ones, including the ones outside the cluster)

 But is cool that we can identify the node that contains the database!

As I see, the basic schema is (please, I am aware that my mental model for HA and Chef is incorrect):

1. Stop openstack pacemaker resources
2. Stop non HA openstack resources
3. Wait (sync) until no openstack service is running in any node of the cloud
4. `pg_dump` and copy the database somewhere else
5. Stop the rest of the pacemaker services, inc, the database, drdb and openais